cmd/fluxctl/install_cmd.go

```diff
@@ -25,49 +29,87 @@ func (opts *installOpts) Command() *cobra.Command {
-		"tell flux it has readonly access to the repo")
-	cmd.Flags().BoolVar(&opts.ManifestGeneration, "manifest-generation", false,
-		"whether to enable manifest generation")
-	cmd.Flags().StringVar(&opts.Namespace, "namespace", "",
-		"cluster namespace where to install flux")
-	cmd.Flags().BoolVar(&opts.RegistryScanning, "registry-scanning", true,
-		"scan container image registries to fill in the registry cache")
-	cmd.Flags().StringVarP(&opts.outputDir, "output-dir", "o", "", "a directory in which to write individual manifests, rather than printing to stdout")
+	cmd.Flags().StringVar(&opts.params.GitUser, "git-user", "Flux", "Username to use as git committer")
+	cmd.Flags().StringVar(&opts.params.GitEmail, "git-email", "", "Email to use as git committer")
+	cmd.Flags().BoolVar(&opts.params.ConfigAsConfigMap, "config-as-configmap", false,
+		"Create a ConfigMap to hold the Flux configuration given with --config-file. If false, a Secret will be used. Ignored if --config-file is not given.")
+	cmd.Flags().BoolVar(&opts.params.GitReadOnly, "git-readonly", false, "Tell flux it has readonly access to the repo")
+	cmd.Flags().BoolVar(&opts.params.ManifestGeneration, "manifest-generation", false, "Whether to enable manifest generation")
+	cmd.Flags().StringVar(&opts.params.Namespace, "namespace", "", "Cluster namespace in which to install Flux")
+	cmd.Flags().BoolVar(&opts.params.RegistryScanning, "registry-scanning", true, "Scan container image registries to fill in the registry cache")
+
+	// These are flags for control of the output, etc.
+	cmd.Flags().StringVar(&opts.configFile, "config-file", "", "Make this file into a secret or configmap for Flux to mount as config")
+	cmd.Flags().StringVarP(&opts.outputDir, "output-dir", "o", "", "A directory in which to write individual manifests, rather than printing to stdout")
 
 	// Hide and deprecate "git-paths", which was wrongly introduced since its inconsistent with fluxd's git-path flag
+	cmd.Flags().StringSliceVar(&opts.params.GitPaths, "git-paths", []string{},
+		"Relative paths within the Git repo for Flux to locate Kubernetes manifests")
 	cmd.Flags().MarkHidden("git-paths")
 	cmd.Flags().MarkDeprecated("git-paths", "please use --git-path (no ending s) instead")
 
 	return cmd
 }
 
 func (opts *installOpts) RunE(cmd *cobra.Command, args []string) error {
-	if opts.GitURL == "" {
-		return fmt.Errorf("please supply a valid --git-url argument")
+	// To make sure the mandatory flags are set, we mimic what fluxd
+	// will do, and load from both the config file (if supplied, it
+	// will be mounted into the fluxd container as a configmap or
+	// secret) and the flags (which will go into the template as fluxd
+	// args).
+
+	if opts.configFile != "" {
+		viper.SetConfigFile(opts.configFile)
+		if err := viper.ReadInConfig(); err != nil {
+			return fmt.Errorf("unable to read config at %s (possibly a missing file extension, try .yaml or .json): %s", opts.configFile, err)
```

Comment:
I would only support a format

Reply:
This is not really supported by viper, so far as I can tell. You can either support all of the formats as selected by the file extension, or specify an exact format and it will look for an extensionless file.

Reply:
I think we can either use a single format (my preference) or, alternatively, do some sort of content detection ourselves and pass the exact expected format to viper.

I don't really see a reason for supporting multiple formats at this point.